doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,229 @@
+- [X] With strict increases for writes following reads: the global timestamp explicitly increases after all writes (
+  See [Group Commit/Write Coalesce](#Group Commit/Write Coalesce)).
+- [X] Each timestamp is assigned to an event at some real-time moment within the event's bounds: For any two events e1
+  and e2, if e2 starts after e1 finished, then e2's timestamp will be larger or equal timestamp than e1's timestamp.
+- [X] Each timestamp transition is made durable before any subsequent response is issued: writes wait for an `append`
+  command to be processed before returning a response to the client.
+    - NOTE: There is ongoing work to ensure that an `append` command doesn't return until it is durable.
+
+### Global Timestamp Recovery
+
+After a restart the Coordinator must reestablish the global timestamp to some value greater than or equal to the value
+of the previous global timestamp before the restart. This ensures that a read or write after the restart is assigned a
+timestamp greater than or equal to all reads and writes before the restart.
+
+For an initial implementation the Coordinator will record all global timestamps in the catalog. On restart the
+Coordinator can read this value from the catalog.
+
+TODO: better implementation.
+
+The properties described in this section and the [Global Timestamp](#Global Timestamp) section are sufficient to ensure
+Strict Serializability across restarts.
+
+### Group Commit/Write Coalesce
+
+Write read cycles (write followed by read followed by write followed by read etc) and consecutive writes can cause the
+global timestamp to increase in an unbounded fashion.
+
+Proposal: All writes across all sessions should be blocked and added to a queue. At the end of X millisecond all pending
+writes are sent in a batch to STORAGE and committed together at the same timestamp. The commits are all assigned a
+timestamp 1 larger than the current global timestamp. Once all commits have been made durable or aborted then the global
+timestamp should be increased to timestamp of the commit. No reads should be served while the writes are committing.
+
+This approach limits the per session write throughput to 1 write transaction per X milliseconds for user tables.
+
+This approach guarantees that when a write completes, the global timestamp is larger than or equal the timestamp of that
+write. Also, when a write completes, all previous writes and reads were at a lower timestamp than the write.
+
+NOTE: If the client had multiple writes to a single table known ahead of time, then grouping them in a single
+multi-statement write transaction would increase throughput. There may be some user education needed for this.
+
+NOTE: This would also fix the problem discussed in [#12198](https://github.com/MaterializeInc/materialize/issues/12198).
+
+### Read Capabilities on Global Timestamp
+
+Proposal: Explicitly hold read capabilities for the global timestamp for all sources. All reads are assigned a timestamp
```

Comment:
Good point. Is it possible to have more than one object per non realtime timelines? If so then I think we actually need one global timestamp per timeline.

Reply:
Yes, there can be multiple objects per timeline. But only the realtime timeline has tables or even a system clock. There's nothing to globally advance on the other timelines because all of the timestamps come exclusively from the source data itself. There's no "clock" on them that moves forward with respect to reality, so I think that most of the stuff in this document doesn't  affect them.

Reply:
I think I need to spend some more time thinking/learning about other timelines. How does compaction work on them if compaction is measured in time but the timelines are not measured in time?

In general I think other timelines would be susceptible to the back in times issue (read from join at `ts`, then read from single object at `< ts`) if the `[since, upper)` ranges for two objects on the same timeline diverge.

Reply:
> How does compaction work on them if compaction is measured in time but the timelines are not measured in time?

We pretend everything is the same units, so 1ms is non-realtime timelines is 1 "whatever" in those timelines too, but users still have to specify the 1 in ms even if it's not ms. Doesn't matter in reality though since no one uses the non-realtime timeline. But since our code knows about them, we do have to care. But yeah, this part is a kinda lie we tell.

> other timelines would be susceptible to the back in times issue

That sounds correct. We might need to define "now" or the "global timestamp" for other timelines as "the oldest since" and then the rest of the doc falls out of that. TBD how to separate out the oldest since but also not comingle it with holding back all the other sinces to that one.

Reply:
I think that defining the "global timestamp" for other timelines as "the oldest since" would work nicely but I still have two open questions I've been trying to think through:
1. I think we would need to hold back the `since`s of all objects on the same timeline to this global timestamp, or else how can we query the other objects at the global timestamp.
2. What advances the global timestamp of non-realtime timelines and how is that communicated to the coordinator?

Reply:
1. I agree. This does get weird, though, if we ack to an upstream source that some write occurred at some time t because of our linearizability goal (if there's another source on that some CDC timeline that has not closed time t yet). These two desires (linearizability/freshness and availability) are in natural conflict with each other.
2. Today there's no concept of global timestamp for other timelines, so nothing. But in general since and upper information is not communicated *to* the Coordinator, instead it can ask dataflow about the current read and write capabilities via the [`collection`](https://dev.materialize.com/api/rust-private/mz_dataflow_types/client/controller/storage/trait.StorageController.html#tymethod.collection) method on the Controller and looking at the `write_frontier` and `implied_capability` properties (https://dev.materialize.com/api/rust-private/mz_dataflow_types/client/controller/storage/struct.CollectionState.html).

Reply:
I think in order to provide Strict Serializability to non-realtime timelines, then we'll need to introduce the concept of a global timestamp per timeline. I think they can behave in the following way:

- It is initialized to `Timestamp::minimum()`
- Periodically a background thread will update the global timestamp s.t. `global_timestamp = max(max(sinces), min(uppers) - 1, global_timestamp)`, where `sinces` is the `since` for all objects in the timeline and `uppers` is the uppers for all objects in the timeline. This thread could even be the same one that advances local inputs every second.

Reply:
Or maybe actually `global_timestamp = max(min(uppers) - 1, global_timestamp)`.

I think the issue with using `since` to advance the global timestamp is that we need to actively hold back `since` so we'll never actually make any progress. I have to think about this for a bit.